src/libopensc/card-epass2003.c

```diff
@@ -1330,6 +1330,100 @@ epass2003_sm_wrap_apdu(struct sc_card *card, struct sc_apdu *plain, struct sc_ap
 	return SC_SUCCESS;
 }
 
+static int
+epass2003_check_response_mac_and_sw(struct sc_card *card, struct sc_apdu *sm)
+{
+	unsigned char iv[16];
+	unsigned char *data = NULL, *mac = NULL;
+	size_t blocksize, mac_len;
+	int ret = -1;
+	size_t taglen;
+	const u8 *tag;
+	epass2003_exdata *exdata;
+	unsigned char *in = sm->resp;
+	unsigned char *alt_in;
+	size_t inlen = sm->resplen;
+
+	/* card/ctx/drv_data is already checked by caller */
+	exdata = (epass2003_exdata *)card->drv_data;
+
+	/* compare BER-TLV encoded SW (TAG 0x99) and raw SW */
+	alt_in = in;
+	tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x99, &taglen);
+	if (tag == NULL || taglen != 2) {
+		/*
+		 * It seems that the EPASS2003 firmware has some problem with BER-TLV encoding.
+		 * There seems to be some proprietary fix for the faulty encoding in the
+		 * decrypt_response() function.  Instead of (correct) TLV 87 81 81 [01 ...]
+		 * incorrect TLV 87 81 [01 ...  ] is returned.
+		 */
+		if (0x01 == in[2] && 0x82 != in[1]) {
+			sc_log(card->ctx, "Workaround, wrong BER-TLV ?");
+			alt_in += in[1] + 2;
+			tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x99, &taglen);
+			if (tag == NULL || taglen != 2)
+				return ret;
+		} else
+			return ret;
+	}
+	if (sm->sw1 != tag[0] || sm->sw2 != tag[1])
+		return ret;
+
+	if (exdata->bFipsCertification) {
+		sc_log(card->ctx, "Warning, MAC is not checked");
+		return 0;
+	}
+	tag = sc_asn1_find_tag(card->ctx, alt_in, inlen, 0x8e, &taglen);
+	if (tag == NULL || taglen != 8)
+		return ret;
+
+	if (KEY_TYPE_AES == exdata->smtype)
+		blocksize = 16;
+	else
+		blocksize = 8;
+
+	mac_len = tag - in - 2;
+	if (NULL == (data = malloc(mac_len + blocksize)))
+		goto end;
+	if (NULL == (mac = malloc(mac_len + blocksize)))
+		goto end;
+
+	/* copy response to buffer and append padding */
+	memcpy(data, in, mac_len);
+	memset(data + mac_len, 0, blocksize);
```

Comment:
just use calloc instead of malloc above and you don't need to memset here

Reply:
It would be best to use `alloca` (let's save `free` maintenance). However, I did not find `alloc' anywhere in the OpenSC code, I understand that there may be a problem with portability.

Reply:
alloca is not supported on Windows, afaik. Also, I don't like it personally because you can't do any error handling (a failure in memory allocation with alloca is just undefined behavior).